File: src/Api/Model/Languageforge/Lexicon/Config/LexConfig.php
<?php

namespace Api\Model\Languageforge\Lexicon\Config;

use Api\Model\Shared\Mapper\ObjectForEncoding;

class LexConfig extends ObjectForEncoding
{
    public function __construct()
    {
        $this->hideIfEmpty = false;
        $this->label = '';
    }

    // config types
    const FIELDLIST = 'fields';
    const MULTITEXT = 'multitext';
    const MULTIPARAGRAPH = 'multiparagraph';
    const OPTIONLIST = 'optionlist';
    const MULTIOPTIONLIST = 'multioptionlist';

    // fields
    const LEXEME = 'lexeme';
    const DEFINITION = 'definition';
    const GLOSS = 'gloss';
    const POS = 'partOfSpeech';
    const PICTURES = 'pictures';
    const SEMDOM = 'semanticDomain';
    const EXAMPLE_SENTENCE = 'sentence';
    const EXAMPLE_TRANSLATION = 'translation';

    // less common FLEx fields
    const CITATIONFORM = 'citationForm';
    const ENVIRONMENTS = 'environments';
    const PRONUNCIATION = 'pronunciation';
    const CVPATTERN = 'cvPattern';
    const TONE = 'tone';
    const LOCATION = 'location';
    const ETYMOLOGY = 'etymology';
    const ETYMOLOGYGLOSS = 'etymologyGloss';
    const ETYMOLOGYCOMMENT = 'etymologyComment';
    const ETYMOLOGYSOURCE = 'etymologySource';
    const NOTE = 'note';
    const LITERALMEANING = 'literalMeaning';
    const ENTRYBIBLIOGRAPHY = 'entryBibliography';
    const ENTRYRESTRICTIONS = 'entryRestrictions';
    const SUMMARYDEFINITION = 'summaryDefinition';
    const ENTRYIMPORTRESIDUE = 'entryImportResidue';

    const SCIENTIFICNAME = 'scientificName';
    const ANTHROPOLOGYNOTE = 'anthropologyNote';
    const SENSEBIBLIOGRAPHY = 'senseBibliography';
    const DISCOURSENOTE = 'discourseNote';
    const ENCYCLOPEDICNOTE = 'encyclopedicNote';
    const GENERALNOTE ='generalNote';
    const GRAMMARNOTE = 'grammarNote';
    const PHONOLOGYNOTE = 'phonologyNote';
    const SENSERESTRCTIONS = 'senseRestrictions';
    const SEMANTICSNOTE = 'semanticsNote';
    const SOCIOLINGUISTICSNOTE = 'sociolinguisticsNote';
    const SOURCE = 'source';
    const USAGES = 'usages';
    const REVERSALENTRIES = 'reversalEntries';
    const SENSETYPE = 'senseType';
    const ACADEMICDOMAINS = 'academicDomains';
    const ANTHROPOLOGYCATEGORIES = 'anthropologyCategories';
    const SENSEIMPORTRESIDUE = 'senseImportResidue';
    const STATUS = 'status';


    const REFERENCE = 'reference';

    // field lists
    const SENSES_LIST = 'senses';
    const EXAMPLES_LIST = 'examples';
    //const CUSTOM_FIELDS_LIST = 'customFields';


    // comments
    const COMMENTS_LIST = 'comments';
    const REPLIES_LIST = 'replies';

    /** @var string */
    public $type;

    /** @var string */
    public $label;
    
    /** @var boolean */
    public $hideIfEmpty;

    /**
     * Index of field names (key) against FLEx option list codes (value)
     * If the FLEx code is not known the field name is used
     *
     * @var array
     */
    private static $flexOptionlistCodes = array(
        self::POS => 'grammatical-info',
        self::SEMDOM => 'semantic-domain-ddp4',
        self::ENVIRONMENTS => self::ENVIRONMENTS,
        self::LOCATION => 'location',
        self::USAGES => 'usage-type',
        self::REVERSALENTRIES => 'reversal-type',
        self::SENSETYPE => 'sense-type',
        self::ACADEMICDOMAINS => 'domain-type',
        self::ANTHROPOLOGYCATEGORIES => 'anthro-code',
        self::STATUS => 'status',
    );

    /**
     * Returns the FLEx option list code given the field name
     * If the code is unknown the field name is returned
     *
     * @param string $fieldName
     * @return string
    */
    public static function flexOptionlistCode($fieldName) {
        if (array_key_exists($fieldName, self::$flexOptionlistCodes)) {
            return self::$flexOptionlistCodes[$fieldName];
        }
        return $fieldName;
    }

    /**
     * Index of FLEx option list codes (key) against option list names (value)
     *
     * @var array
     */
    private static $flexOptionlistNames = array(
        'grammatical-info' => 'Part of Speech',
        'semantic-domain-ddp4' => 'Semantic Domain',
        'domain-type' => 'Academic Domains',
        self::ENVIRONMENTS => 'Environments',
        'location' => 'Location',
        'usage-type' => 'Usages',
        'reversal-type' => 'Reversal Entries',
        'sense-type' => 'Type',
        'anthro-code' => 'Anthropology Categories',
        'do-not-publish-in' => 'Publish In',
        'status' => 'Status',

        'etymology' => 'Etymology',
        'lexical-relation' => 'Lexical Relation',
        'note-type' => 'Note Type',
        'paradigm' => 'Paradigm',
        'users' => 'Users',
        'translation-type' => 'Translation Type',
        'from-part-of-speech' => 'From Part of Speech',
        'morph-type' => 'Morph Type',
        'noun-slot' => 'Noun Slot',
        'verb-slot' => 'Verb Slot',
        'stative-slot' => 'Stative Slot',
        'noun-infl-class' => 'Noun Inflection Class',
        'verb-infl-class' => 'Verb Inflection Class'
    );

    /**
     * Returns the option list name given the FLEx option list code
     * If the name is unknown the code is returned
     *
     * @param string $listCode
     * @return string
     */
    public static function flexOptionlistName($listCode) {
        if (array_key_exists($listCode, self::$flexOptionlistNames)) {
            return self::$flexOptionlistNames[$listCode];
        }
        return $listCode;
    }
}

File: src/netcore-api/SIL.XForge.WebApi.Server/Models/Lexicon/LexConfig.cs
namespace SIL.XForge.WebApi.Server.Models.Lexicon
{
    public abstract class LexConfig
    {
        // config types
        public const string FieldList = "fields";
        public const string MultiText = "multitext";
        public const string MultiParagraph = "multiparagraph";
        public const string OptionList = "optionlist";
        public const string MultiOptionList = "multioptionlist";

        // fields
        public const string Lexeme = "lexeme";
        public const string Definition = "definition";
        public const string Gloss = "gloss";
        public const string Pos = "partOfSpeech";
        public const string Pictures = "pictures";
        public const string Semdom = "semanticDomain";
        public const string ExampleSentence = "sentence";
        public const string ExampleTranslation = "translation";

        // less common FLEx fields
        public const string CitationForm = "citationForm";
        public const string Environments = "environments";
        public const string Pronunciation = "pronunciation";
        public const string CVPattern = "cvPattern";
        public const string Tone = "tone";
        public const string Location = "location";
        public const string Etymology = "etymology";
        public const string EtymologyGloss = "etymologyGloss";
        public const string EtymologyComment = "etymologyComment";
        public const string EtymologySource = "etymologySource";
        public const string Note = "note";
        public const string LiteralMeaning = "literalMeaning";
        public const string EntryBibliography = "entryBibliography";
        public const string EntryRestrictions = "entryRestrictions";
        public const string SummaryDefinition = "summaryDefinition";
        public const string EntryImportResidue = "entryImportResidue";

        public const string ScientificName = "scientificName";
        public const string AnthropologyNote = "anthropologyNote";
        public const string SenseBibliography = "senseBibliography";
        public const string DiscourseNote = "discourseNote";
        public const string EncyclopedicNote = "encyclopedicNote";
        public const string GeneralNote ="generalNote";
        public const string GrammarNote = "grammarNote";
        public const string PhonologyNote = "phonologyNote";
        public const string SenseRestrictions = "senseRestrictions";
        public const string SemanticsNote = "semanticsNote";
        public const string SociolinguisticsNote = "sociolinguisticsNote";
        public const string Source = "source";
        public const string Usages = "usages";
        public const string ReversalEntries = "reversalEntries";
        public const string SenseType = "senseType";
        public const string AcademicDomains = "academicDomains";
        public const string AnthropologyCategories = "anthropologyCategories";
        public const string SenseImportResidue = "senseImportResidue";
        public const string Status = "status";

        public const string Reference = "reference";

        // field lists
        public const string SensesList = "senses";
        public const string ExamplesList = "examples";
        //public const string CustomFieldsList = "customFields";

        // comments
        public const string CommentsList = "comments";
        public const string RepliesList = "replies";

        public string Type { get; protected set; }
        public string Label { get; set; }
        public bool HideIfEmpty { get; set; }

    }
}
